### ISSUE TITLE
requirements not installed with the plugin

### ISSUE THREAD

**Issue by ttomasz** [kauevestena/osm_sidewalkreator#6]
I'm using Flatpak version. I tried to install plugin but got error that module osm2geojson or shapely is not installed.
Tried reinstalling plugin and restarting qgis but it doesn't help.

**Comment by kauevestena** [comment]
Hello Thomasz!!

Sadly, shapely is a dependency of osm2geojson, I will try to implement ogr2ogr in the code in order to remove even this tiny dependency...

Try to install shapely on your QGIS python, I don't know for sure if a python inside flatpak can be accessed directly, but try to:

     <qgis_python_path> -m pip install shapely

**Comment by kauevestena** [comment]
I think that for flatpak, things get more complicated: 

https://stackoverflow.com/questions/68077519/install-extra-packages-in-flatpak-sandbox

**Comment by kauevestena** [comment]
This one seems more promissing:

https://github.com/flatpak/flatpak-builder-tools/tree/master/pip

https://github.com/flatpak/flatpak-builder-tools/tree/master/pip

**Comment by ttomasz** [comment]
I tried earlier with pip install but it did not recognize pip and tried to run ensure_pip but it was not able to install anything.

If the plugin only downloads specific things and the geometries are not too complicated we could do some simple parsing ourselves probably. I can try to help with that.

**Comment by kauevestena** [comment]
Thank you man!!

I will try to replace the call for osm2geojson (yeah it's a single one), for a call of

    gdal.VectorTranslate()

like here: https://gis.stackexchange.com/a/418872/49900

**Comment by kauevestena** [comment]
hey @ttomasz , can you test running the following commands in your QGIS bundled python console:

    import subprocess, sys
    subprocess.check_call([sys.executable, "-m", "pip", "install", 'shapely'])

I can bundle this alongside the code!
(actually I tried it for osm2geojso, it is as a comment at line 107 in osm_sidewalkreator.py)

**Comment by ttomasz** [comment]
```
import subprocess, sys
subprocess.check_call([sys.executable, "-m", "pip", "install", 'shapely'])
Traceback (most recent call last):
  File "/usr/lib/python3.9/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/usr/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', '-m', 'pip', 'install', 'shapely']' returned non-zero exit status 1.
```
i don't think there is pip module at all in flatpak but i'm not sure

**Comment by ttomasz** [comment]
Have you tried creating temporary files using https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile ? it should be multiplatform. You can get file path after initializing object by `.name` attribute.

**Comment by kauevestena** [comment]
> Have you tried creating temporary files using https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile ? it should be multiplatform. You can get file path after initializing object by `.name` attribute.

never tried, sounds good!! 
just kinda lazy to boot on windows to test, but I will!!

**Comment by kauevestena** [comment]
> ```
> import subprocess, sys
> subprocess.check_call([sys.executable, "-m", "pip", "install", 'shapely'])
> Traceback (most recent call last):
>   File "/usr/lib/python3.9/code.py", line 90, in runcode
>     exec(code, self.locals)
>   File "<input>", line 1, in <module>
>   File "/usr/lib/python3.9/subprocess.py", line 373, in check_call
>     raise CalledProcessError(retcode, cmd)
> subprocess.CalledProcessError: Command '['/usr/bin/python3', '-m', 'pip', 'install', 'shapely']' returned non-zero exit status 1.
> ```
> 
> i don't think there is pip module at all in flatpak but i'm not sure

maybe...

to check you may try

    import subprocess, sys
    subprocess.check_call([sys.executable, "-m", "ensurepip", -- upgrade'])

or 

    import subprocess, sys
    subprocess.check_call([sys.executable, "get-pip.py"'])

adapted from https://pip.pypa.io/en/stable/installation/#python

**Comment by waldyrious** [comment]
I'm facing the same error. Here's a screenshot in case it helps others confirm they have the same issue:

![image](https://github.com/user-attachments/assets/8d93182e-ee7d-4a39-926c-58e396031579)

I tried the suggestions from the comment above, but got the following:

```python
import subprocess, sys
subprocess.check_call([sys.executable, "-m", "ensurepip", "-- upgrade"])
Traceback (most recent call last):
  File "/usr/lib/python3.9/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/usr/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', '-m', 'ensurepip', '-- upgrade']' returned non-zero exit status 2.
```

```python
subprocess.check_call([sys.executable, "get-pip.py"])
Traceback (most recent call last):
  File "/usr/lib/python3.9/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/usr/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', 'get-pip.py']' returned non-zero exit status 2.
```

(Note that in both cases I had to fix the quoting of the `check_call` arguments because there was a stray single quote at the end.)

Any hints at how to proceed?

**Comment by kauevestena** [comment]
Hello @waldyrious , first of all, thanks a lot for your interest on the plugin!

Well, I made a [branch ](https://github.com/kauevestena/osm_sidewalkreator/tree/remove_dependencies) aimed on fully removing the single dependency of the plugin, that even comes bundled. But on its end, it needs shapely, which depends on a bigger ecosystem.  So I plan to in the next weeks accomplish this long lasting TODO.

Going back to the present state, are you using flatpak?

**Comment by kauevestena** [comment]
In case of flatpat, try this steps:

- Access the Flatpak Shell

    flatpak run --command=sh org.qgis.qgis
    
  - Install pip manually:
  
      curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
      python3 get-pip.py
      
  - Install shapely
  
      pip install shapely
      
      
  Other methods for installing pip can be used, seems that accessing Flatpak shell is key here!

**Comment by waldyrious** [comment]
Yeah, I am indeed using Flatpak.

I tried running the commands you suggested, but they seem to fail silently. I ran `flatpak run --command=sh org.qgis.qgis`, and then in the resulting session, typed the commands you suggested, which didn't give out any errors, but then failed when I tried to actually use pip:

```
[📦 org.qgis.qgis ~]$ curl bootstrap.pypa.io/get-pip.py -o get-pip.py
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[📦 org.qgis.qgis ~]$ python3 get-pip.py
[📦 org.qgis.qgis ~]$ pip install shapely
sh: pip: command not found
[📦 org.qgis.qgis ~]$ 
```

However, I managed to get it to work with the [standalone zip application](https://pip.pypa.io/en/stable/installation/#standalone-zip-application) installation method mentioned in the pip documentation:

```
[📦 org.qgis.qgis ~]$ curl https://bootstrap.pypa.io/pip/pip.pyz -o pip.pyz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1745k  100 1745k    0     0  2723k      0 --:--:-- --:--:-- --:--:-- 2723k
[📦 org.qgis.qgis ~]$ python3 pip.pyz install shapely
Defaulting to user installation because normal site-packages is not writeable
Collecting shapely
  Downloading shapely-2.0.5-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (7.0 kB)
Requirement already satisfied: numpy<3,>=1.14 in /app/lib/python3.9/site-packages (from shapely) (1.23.4)
Downloading shapely-2.0.5-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.5 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 7.4 MB/s eta 0:00:00
Installing collected packages: shapely
Successfully installed shapely-2.0.5
```

With this, I was able to finally install the plugin without getting any error message! :tada::tada::tada:
  
**TL;DR** just to make it easier to follow, for those who installed QGIS using Flatpak, the commands to run are:

- Open the flatpak shell for the qgis package: `flatpak run --command=sh org.qgis.qgis`
- Within the shell, type:
  - `curl https://bootstrap.pypa.io/pip/pip.pyz -o pip.pyz`
  - `python3 pip.pyz install shapely`

Hope this helps!

**Comment by kauevestena** [comment]
Thanks a lot for finding the workaround for Flatpak users @waldyrious !! 

It will be added to the readme file!

If you have any other issue or doubt, please don't hesitate to ask! 


Um grande abraço, meu caro!
